cwe_checker_rs/src/checkers/cwe_78/context/mod.rs

```diff
@@ -0,0 +1,563 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use petgraph::{graph::NodeIndex, visit::IntoNodeReferences};
+
+use super::{state::State, taint::Taint, CWE_MODULE};
+use crate::{
+    abstract_domain::AbstractDomain,
+    analysis::{
+        forward_interprocedural_fixpoint::Context as PiContext,
+        graph::{Graph, Node},
+        interprocedural_fixpoint_generic::NodeValue,
+        pointer_inference::PointerInference as PointerInferenceComputation,
+        pointer_inference::State as PointerInferenceState,
+    },
+    bil::Bitvector,
+    intermediate_representation::*,
+    utils::log::CweWarning,
+};
+
+#[derive(Clone)]
+pub struct Context<'a> {
+    /// A pointer to the corresponding project struct.
+    project: &'a Project,
+    /// The reversed pointer inference graph
+    graph: Graph<'a>,
+    /// A pointer to the results of the pointer inference analysis.
+    /// They are used to determine the targets of pointers to memory,
+    /// which in turn is used to keep track of taint on the stack or on the heap.
+    pub pointer_inference_results: &'a PointerInferenceComputation<'a>,
+    /// A map to get the node index of the `BlkStart` node containing a given [`Def`] as the last `Def` of the block.
+    /// The keys are of the form `(Def-TID, Current-Sub-TID)`
+    /// to distinguish the nodes for blocks contained in more than one function.
+    block_start_node_map: Arc<HashMap<(Tid, Tid), NodeIndex>>,
+    /// Maps the TID of an extern symbol to the extern symbol struct.
+    extern_symbol_map: Arc<HashMap<Tid, &'a ExternSymbol>>,
+    /// Maps the TID of an extern string related symbol to the corresponding extern symbol struct.
+    string_symbol_map: Arc<HashMap<Tid, &'a ExternSymbol>>,
+    /// A map to get the node index of the `BlkEnd` node containing a given [`Jmp`].
+    /// The keys are of the form `(Jmp-TID, Current-Sub-TID)`
+    /// to distinguish the nodes for blocks contained in more than one function.
+    jmp_to_blk_end_node_map: Arc<HashMap<(Tid, Tid), NodeIndex>>,
+    /// The call whose parameter values are the sources for taint for the analysis.
+    pub taint_source: Option<&'a Term<Jmp>>,
+    /// The name of the function, whose parameter values are the taint sources.
+    pub taint_source_name: Option<String>,
+    /// The current subfunction.
+    ///Since the analysis is intraprocedural,
+    ///all nodes with state during the fixpoint algorithm should belong to this function.
+    current_sub: Option<&'a Term<Sub>>,
```

Comment:
Is the analysis really intraprocedural? You do propagate taint across edges not belonging to your subfunction if I read the context trait implementation correctly.

Reply:
Oops. Must have misread that when I copied the comment.